spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BCECriterion.scala

```diff
@@ -37,48 +37,96 @@ import scala.reflect.ClassTag
+    require(input.size().sameElements(target.size()),
+      s"input size should be equal to target size, but got input size: ${input.size().toList}," +
+        s" target size: ${input.size().toList}")
+
+    if (weights != null) {
+      if (weights.nDimension() < input.nDimension()) {
+        require(weights.size().sameElements(input.size().tail),
+          s"weights size should be equal to input size or input size's tail, but got" +
+            s" input size: ${input.size().toList}, weights size: ${weights.size().toList}")
+      } else if (weights.nDimension() == input.nDimension()) {
+        require(weights.size().sameElements(input.size()),
+          s"weights size should be equal to input size or input size's tail, but got" +
+            s" input size: ${input.size().toList}, weights size: ${weights.size().toList}")
+      } else {
+        throw new IllegalArgumentException(
+          s"weights size should be equal to input size or input size's tail, but got" +
+          s" input size: ${input.size().toList}, weights size: ${weights.size().toList}")
+      }
     }
 
+    val nElement = input.nElement()
+
     var sum = 0.0
     if (null != weights) {
-      val func = new TensorFunc6[T] {
-        override def apply(data1: Array[T], offset1: Int, data2: Array[T], offset2: Int,
-                           data3: Array[T], offset3: Int): Unit = {
-          val x = ev.toType[Double](data1(offset1))
-          val y = ev.toType[Double](data2(offset2))
-          val w = ev.toType[Double](data3(offset3))
-          sum -= (Math.log(x + eps) * y + Math.log(1.0 - x + eps) * (1.0 - y)) * w
+      if (weights.nElement() < 200) {
+        // simple heuristic, if the input tensor is small enough, the
+        // vectorized implementation is actually slower
+
+        if (target.dim() > weights.dim()) {
+          val size = target.size()
+          size(0) = 1
+          expendedWeights = weights.view(size).expandAs(target)
+        }
+
+        val func = new TensorFunc6[T] {
+          override def apply(data1: Array[T], offset1: Int, data2: Array[T], offset2: Int,
+                             data3: Array[T], offset3: Int): Unit = {
+            val x = ev.toType[Double](data1(offset1))
+            val y = ev.toType[Double](data2(offset2))
+            val w = ev.toType[Double](data3(offset3))
+            sum += (Math.log(x + eps) * y + Math.log(1.0 - x + eps) * (1.0 - y)) * w
+          }
         }
+        DenseTensorApply.apply3(input, target, expendedWeights, func)
+      } else {
+        buffer.resizeAs(input).copy(input).add(ev.fromType(eps)).log()
+        // cmul support broadcasting
+        buffer.cmul(weights)
+        sum += ev.toType[Double](buffer.dot(target))
+        buffer.fill(ev.fromType(1.0 + eps)).sub(input).log().cmul(weights)
+        sum -= ev.toType[Double](buffer.dot(target))
+        onesBuffer.resizeAs(buffer).fill(ev.fromType(1.0))
```

Comment:
add `if onesBuffer.nElement() == buffer.nElement()` to decide whether resize the `onesBuffer`

Reply:
@qiuxin2012 done